### ISSUE TITLE
Constrain class decorators with new()

### ISSUE THREAD

**Issue by Ciantic** [Microsoft/TypeScript#9192]
Following code does not give any errors at the moment:
```typescript
class Bar {

}

function decorator<T extends Bar>(constructor: new () => T) {
    return constructor;
};


@decorator
class Foo extends Bar {
}

@decorator // I want this to be error!
class NotAllowed {
}
```

P.S. if my above example does not make sense, what I want is to set constrains on which classes the decorator can be used.

**Comment by aluanhaddad** [comment]
The issues is that because `Bar` has no members it is structurally compatible with `NotAllowed`.
The following produces an error where you have indicated it should.
```TypeScript
class Bar {
    value: string = "abc"

}

function decorator<T extends Bar>(constructor: new () => T) {
    return constructor;
};


@decorator
class Foo extends Bar {
}

@decorator // I want this to be error!
class NotAllowed {
}
```

**Comment by aluanhaddad** [comment]
![image](https://cloud.githubusercontent.com/assets/5034656/16095342/286d908e-3312-11e6-8775-f57ed87ec213.png)
![image](https://cloud.githubusercontent.com/assets/5034656/16095355/3380b0be-3312-11e6-8c7b-9e463688c2bc.png)

**Comment by Ciantic** [comment]
I was struck by structural typing here! Thanks.

**Comment by aluanhaddad** [comment]
Glad I could help!
